src/cargo/ops/registry.rs

```diff
@@ -943,16 +901,41 @@ pub fn yank(
 /// Gets the SourceId for an index or registry setting.
 ///
 /// The `index` and `reg` values are from the command-line or config settings.
-/// If both are None, returns the source for crates.io.
-fn get_source_id(config: &Config, index: Option<&str>, reg: Option<&str>) -> CargoResult<SourceId> {
-    match (reg, index) {
-        (Some(r), _) => SourceId::alt_registry(config, r),
-        (_, Some(i)) => SourceId::for_registry(&i.into_url()?),
-        _ => {
-            let map = SourceConfigMap::new(config)?;
-            let src = map.load(SourceId::crates_io(config)?, &HashSet::new())?;
-            Ok(src.replaced_source_id())
+/// If both are None, and no source-replacement is configured, returns the source for crates.io.
+/// If both are None, and source replacement is configured, returns an error (except when testing Cargo).
+///
+/// If `reg` is set, source replacement is not followed.
+fn get_source_id(
+    config: &Config,
+    index: Option<&str>,
+    reg: Option<&str>,
+) -> CargoResult<(SourceId, SourceId)> {
+    let map = SourceConfigMap::new(config)?;
+    let sid = match (reg, index) {
+        (None, None) | (Some("crates-io"), None) => SourceId::crates_io(config)?,
+        (Some(r), None) => SourceId::alt_registry(config, r)?,
+        (None, Some(i)) => SourceId::for_registry(&i.into_url()?)?,
+        (Some(_), Some(_)) => {
+            bail!("both `--index` and `--registry` should not be set at the same time")
         }
+    };
+    let src = map.load(sid, &HashSet::new())?;
+    if src.is_replaced() && reg.is_none() && index.is_none() {
+        // Neither --registry nor --index was passed and source replacement is configured.
+        let replacement_sid = src.replaced_source_id();
+        if replacement_sid.is_default_registry()
+            || std::env::var("__CARGO_TEST_ALLOW_CRATES_IO_REPLACEMENT_DO_NOT_USE").is_ok()
+        {
+            // Allow replacement of crates.io for the sparse crates.io endpoint or for Cargo's
+            // testing framework.
+            Ok((replacement_sid, sid))
+        } else if let Some(replacement_name) = replacement_sid.alt_registry_key() {
+            bail!("crates-io is replaced with remote registry {replacement_name};\ninclude `--registry {replacement_name}` or `--registry crates-io`");
+        } else {
+            bail!("crates-io is replaced with non-remote-registry source {replacement_sid};\ninclude `--registry crates-io` to publish to crates.io");
```

Comment:
Are there tests that hit each of these code paths? And also tests that hit each of the suggestions?

Reply:
Both of these do have tests in `publish.rs`: `publish_with_bad_source` and `cratesio_source_replacement`.

The suggestion to use `--registry dummy-registry` has extensive coverage (almost all publishing tests use it).
The `--registry crates-io` suggestion has no coverage, since it would mean publishing to real crates.io.